### PR TITLE
add InvoiceItem class

### PR DESCRIPTION
**Pull request by MaciejBesaraba**
Implement InvoiceItem class data model, please review.

### CODE REVIEW

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 1]
```diff
@@ -1,14 +1,130 @@
...
     {
-        // TODO MaBa
+        public IObjectIdentifier<ulong> Id { get; }
+        public string Name { get; }
+        public decimal UnitPrice { get; }
+        public EUnitType UnitType { get; }
+        public decimal Discount { get; }
+        public double Quantity { get; }
+
+
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Code style -> 2 blank lines between auto properties and constructors

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 2]
```diff
@@ -1,14 +1,130 @@
...
+        public InvoiceItem(
+            IObjectIdentifier<ulong> id,
+            string name,
+            decimal unitPrice,
+            EUnitType unitType,
+            decimal discount,
+            double quantity
+        )
+
+
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Code style -> delete blank lines

```c#
public InvoiceItem(
    IObjectIdentifier<ulong> id,
    string name,
    decimal unitPrice,
    EUnitType unitType,
    decimal discount,
    double quantity
)
{
    Id = id;
    Name = name;
    UnitPrice = unitPrice;
    UnitType = unitType;
    Discount = discount;
    Quantity = quantity;
}
```

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 3]
```diff
@@ -1,14 +1,130 @@
...
+            Id = id;
+            Name = name;
+            UnitPrice = unitPrice;
+            UnitType = unitType;
+            Discount = discount;
+            Quantity = quantity;
+
+        }
+
+
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs:44]: Code style -> 2 blank lines between constructors and properties

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 4]
```diff
@@ -1,14 +1,130 @@
...
+
+        public decimal Total => UnitPrice - Discount;
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: `Discount` is considered as full discount on all items undependent of `UnitPrice` and `Quanitity`.
`Total` is a full price payed for aggregated value of this item set. You should include `Quantity` here.

Item of price 2.50 with quantity of 2 and no discount should result in Total = 5.00
Item of price 2.50 with quantity of 2 and discount of 1 should result in Total = 4.00

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 5]
```diff
@@ -1,14 +1,130 @@
...
+
+        public decimal Subtotal => Total * Quantity;
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Subtotal is a `Total` (above conversation), but it does not include `Discount`.

Item of price 2.50 with quantity of 2 and no discount should result in Subotal = 5.00
Item of price 2.50 with quantity of 2 and discount of 1 should result in Subotal = 5.00

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 6]
```diff
@@ -1,14 +1,130 @@
...
+             
+
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs:47]: Code style -> 2 blank lines

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 7]
```diff
@@ -1,14 +1,130 @@
...
+        public decimal Total => UnitPrice - Discount;
+
+        public decimal Subtotal => Total * Quantity;
+             
+                
+
+
+        public override string ToString()
+        {
+
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Code style -> no blank lines between "`{`" and "`return`"

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 8]
```diff
@@ -1,14 +1,130 @@
...
+            return "InvoiceItem(" +
+                       $"id={Id}, " +
+                       $"name={Name.ToString()}, " +
+                       $"unitPrice={UnitPrice.ToString(CultureInfo.InvariantCulture)}, " +
+                       $"unitType={UnitType}, " +
+                       $"discount={Discount.ToString(CultureInfo.InvariantCulture)}, " +
+                       $"subtotal={Subtotal.ToString(CultureInfo.InvariantCulture)}, " +
+                       $"total={Total.ToString(CultureInfo.InvariantCulture)}, " +
+                       $"quantity={Quantity.ToString()}, " +
+
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Code style -> delete blank line

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 9]
```diff
@@ -1,14 +1,130 @@
...
+            {
+                return false;
+            }
+
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            // TODO ArBy redemptionDate equality
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: `// TODO ...` is a comment with a tag meaning `to do later` you probaly shoud not leave todo tags for me.
And this one is not valid here so delete it.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 10]
```diff
@@ -1,14 +1,130 @@
...
+            return Equals(Id, other.Id) &&
+                   Equals(Name, other.Name) &&
+                   Equals(UnitPrice, other.UnitPrice) &&
+                   Equals(UnitType, other.UnitType) &&
+                   Equals(Discount, other.Discount) &&
+                   Equals(Quantity, other.Quantity);
+
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Code style -> delete blank line

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs [hunk 11]
```diff
@@ -1,14 +1,130 @@
...
+        {
+            unchecked
+            {
+                var hashCode = (Id != null ? Id.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (Name != null ? Name.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (UnitPrice != null ? UnitPrice.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (UnitType != null ? UnitType.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (Discount != null ? Discount.GetHashCode() : 0);
+                hashCode = (hashCode * 397) ^ (Quantity != null ? Quantity.GetHashCode() : 0);
+
```
Comment (cafetamine) [inline InvoiceManagementStudio/Model/Domain/InvoiceItem.cs:108]: Code style -> single blank line before return statement in bigger batch

**Review by cafetamine — CHANGES_REQUESTED** [review]
1. Please think through `Subtotal` and `Total` properties, added explanatory descriptions. In case I'm wrong write in conversation regarding them.

1. Please fallow code style comments regarding unnecesary blank lines.

1. Next time try to add more commits for your objects. My flow would look somewhat like:
    - add auto properties
    - implement required arguments constructor
    - implement properties
    - implement IEqutable interface
    - implement ToString

    and regarding this review changes:
    - delete unnecessary balnk spaces
    - fix properties Total and Subtotal

1. Remeber that you don't have to create a seperate pull request for result of your work on this review. After implementing neccessary changes just push your branch (`invoiceItem`) to origin.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 1 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Fixed.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 2 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Fixed.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 3 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs:44]: Fixed.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 4 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Not sure if know what you mean. As far as im concerned, Total should include UnitPrice multiplied by Quantity, right?

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 5 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Fixed. Please review.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 7 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Fixed.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 8 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Fixed.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 9 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Sorry, just forget to delete it.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 10 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs]: Fixed.

File: InvoiceManagementStudio/Model/Domain/InvoiceItem.cs (hunk 11 above)
Comment (MaciejBesaraba) [reply InvoiceManagementStudio/Model/Domain/InvoiceItem.cs:108]: Fixed.

**Review by cafetamine — APPROVED** [review]

**Pull request merged**